### PR TITLE
Feature 11796 create a new environment entity and index dto

### LINKED ISSUE #11796: Create a new Environment entity and IndexDTO [5]
### Feature Description

As the first step for the environmental surveillance module (#11533), a new entity, DTO, and a few accompanying enums need to be created in order to be used in the frontend. As the frontend will be built in Angular, this issue has been split from #11535 which only contains the UI part.

### Added Value/Benefit

This is the first step towards empowering SORMAS for one health.

### Acceptance Criteria

- [x] A new `Environment` entity (extending CoreAdo) has been added with the following fields:
  - Date `reportDate` (Report date)
  - User `reportingUser` (Reporting user)
  - String `environmentName` (Environment name) - database type `text`
  - String `description` (Description) - database type `text`
  - String `externalId` (External ID) - database type `varchar(255)`
  - User `responsibleUser` (Responsible user)
  - InvestigationStatus `investigationStatus` (Investigation status)
  - EnvironmentMedia `environmentMeda` (Environment media)
  - WaterType `waterType` (Water type)
  - String `otherWaterType` - database type `text`
  - EnvironmentInfrastructureDetails `infrastructureDetails` (Infrastructure details)
  - String `otherInfrastructureDetails` - database type `text`
  - Map<WaterUse, Boolean> `waterUse` (see Event.epidemiologicalEvidenceDetails for comparison)
  - String `otherWaterUse` - database type `text`
  - Location `location` (Location)
- [x] A corresponding DTO and ReferenceDto have been created, along with a facade (extending CoreFacade) and service
- [x] An `EnvironmentIndexDto` has been added with the following fields:
  - UUID
  - External ID
  - Environment name
  - Environment media
  - Region (taken from location)
  - District (taken from location)
  - Community (taken from location)
  - GPS lat
  - GPS lon
  - Postal code
  - City
  - Report date
  - Investigation status
- [x] Successful creation of the `EnvironmentIndexDto` is covered by a unit test
- [x] An `EnvironmentCriteria` class has been added, alongside a `buildCriteriaFilter` method in the service, that supports the following filters:
  - freeTextFilter - filters by uuid, externalId, environmentName, description
  - region
  - district
  - community
  - reportDateFrom & reportDateTo - date range by date or by epi week, e.g. similar to the case directory, but without the reference date field because we always use reportDate as the reference
  - investigationStatus
  - environmentMedia
  - relevanceStatus (EntityRelevanceStatus)
  - responsibleUser
  - gpsLatFrom, gpsLatTo, gpsLonFrom, gpsLonTo - filters by environments for which the GPS coordinates fall into the specified range, basically same logic as for dates; if only lat or only lon is specified, the other is ignored; if only from or to is specified, everything above from or below to counts
- [x] A new `WaterType` enum has been added with the following values (trivial captions ommitted):
  - `WASTEWATER`
  - `GROUNDWATER`
  - `SURFACE_WATER` - Surface water (lakes, rivers, runoff, etc.)
  - `PRECIPITATION` - Precipitation (rain or snow)
  - `OTHER`
  - `UNKNOWN`
- [x] A new `EnvironmentInfrastructureDetails` enum has been added with the following values:
  - `SEPTIC_TANK`
  - `LATRIN`
  - `TOILET`
  - `MANHOLE`
  - `WELLS`
  - `SURFACE_WATER`
  - `OPEN_DRAIN`
  - `OTHER`
  - `UNKNOWN`
- [x] A new `EnvironmentMedia` enum has been added with the following values:
  - `WATER`
  - `SOIL_ROCK` - Soil or rock
  - `AIR`
  - `BIOTA`
- [x] A new `WaterUse` enum has been added with the following values:
  - DRINKING_HOUSEHOLD_NEEDS - Drinking and household needs
  - RECREATION
  - INDUSTRY_COMMERCE - Industry and commerce
  - AGRICULTURE - Agriculture (plants or livestock)
  - THERMOELECTRICITY_ENERGY - Thermoelectricity/Energy
  - OTHER - Other
  - UNKNOWN - Unknown
- [x] i18n has been added for all new fields/enum values
- [x] Synchronization logic between the web and mobile app has been added
- [x] The `createUserFilter` method in the service has been implemented; users have access to environments based on their jurisdiction and whether or not they're the reporting user

### Implementation Details

_No response_

### Mockups

_No response_

### Additional Information

The criteria needs to be extended by a `disease` field later, but since this depends on the samples associated with the environment entity, this can't be done right now.

### PR DESCRIPTION
**Pull request by ChristopherRiedel**
Fixes #11796

### CODE REVIEW

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
@ChristopherRiedel Test failures :)

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12076

File: sormas-app/app/src/main/java/de/symeda/sormas/app/backend/caze/Case.java [hunk 1]
```diff
@@ -76,7 +76,8 @@
 
 @Entity(name = Case.TABLE_NAME)
 @DatabaseTable(tableName = Case.TABLE_NAME)
-public class Case extends PseudonymizableAdo {
+public class
```
Comment (MateStrysewske) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/backend/caze/Case.java:79]: I guess this is an unintended change

File: sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentFacadeEjb.java [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+		List<EnvironmentIndexDto> environments = new ArrayList<>();
+
+		IterableHelper.executeBatched(indexListIds, ModelConstants.PARAMETER_LIMIT, batchedIds -> {
+			CriteriaBuilder cb = em.getCriteriaBuilder();
+			CriteriaQuery<EnvironmentIndexDto> cq = cb.createQuery(EnvironmentIndexDto.class);
+			Root<Environment> environment = cq.from(Environment.class);
+
+			final EnvironmentJoins environmentJoins = new EnvironmentJoins(environment);
+//			final EnvironmentQueryContext environmentQueryContext = new EnvironmentQueryContext(cb, cq, environment, environmentJoins);
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentFacadeEjb.java:119]: Remove this if it's not needed anymore

File: sormas-rest/src/main/java/de/symeda/sormas/rest/exception/mappers/CatchAllExceptionMapper.java [hunk 3]
```diff
@@ -36,6 +36,7 @@ public Response toResponse(Exception exception) {
 				+ "If the mapper exists, make sure the exception class is declared as a javax.ejb.ApplicationException.",
 			exception.getClass().getName());
 		String message = exception.getLocalizedMessage();
+		exception.printStackTrace();
```
Comment (MateStrysewske) [inline sormas-rest/src/main/java/de/symeda/sormas/rest/exception/mappers/CatchAllExceptionMapper.java:39]: Remove this line (was probably intended for testing?)

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
Please change the database types and length constraints of all fields that I've specified as using the datatype "text" in the issue description (i.e. use text instead of varchar, and CHARACTER_LIMIT_TEXT)

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12076

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
Data type changes should also be applied to the Android app

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12076

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
Character limits need to be changed in the app entity class as well

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12076

**Review by MateStrysewske — APPROVED** [review]

**Pull request merged**